### PR TITLE
Move evaluation of resource and operation expressions to libcrmcommon

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

File: lib/common/tests/rules/pcmk__evaluate_op_expression_test.c [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    assert_op_expression(EXPR_NAME_MISSING, pcmk_rc_unpack_error);
+}
+
+#define EXPR_ALL_MATCH                                      \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "          \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+reference_name_missing(void **state)
```
Comment (nrwahl2) [inline lib/common/tests/rules/pcmk__evaluate_op_expression_test.c:109]: May want a test for both names missing to ensure that `NULL == NULL` doesn't cause the expression to pass in the future.

File: lib/common/tests/rules/pcmk__evaluate_op_expression_test.c [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                     reference_rc);
+    free_xml(xml);
+}
+
+
+/*
+ * Invalid arguments
+ */
+
+#define EXPR_NO_MATCH                                       \
```
Comment (nrwahl2) [inline lib/common/tests/rules/pcmk__evaluate_op_expression_test.c:52]: We don't meaningfully use this. The only use is in a case where `rule_input == NULL`.

The name `EXPR_NO_MATCH` suggests that we care about mismatches in both the name and the interval, but we never actually check either one.

Could just drop this and use a different XML string constant in `null_invalid()` if you want

File: lib/common/tests/rules/pcmk__evaluate_op_expression_test.c [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+
+/*
+ * Test PCMK_XA_INTERVAL
+ */
+
+#define EXPR_INVALID_INTERVAL                               \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "          \
+        PCMK_XA_INTERVAL "='not-an-interval' />"
```
Comment (nrwahl2) [inline lib/common/tests/rules/pcmk__evaluate_op_expression_test.c:135]: empty string is also an invalid interval; up to you whether to test it separately or leave it alone

File: lib/common/rules.c [hunk 4]
```diff
@@ -1155,41 +1155,66 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
+    value = crm_element_value(rsc_expression, PCMK_XA_TYPE);
+    if (!pcmk__str_empty(value)
+        && !pcmk__str_eq(value, rule_input->rsc_agent, pcmk__str_none)) {
+        crm_trace(PCMK_XE_RSC_EXPRESSION " %s fails because "
+                  "agent doesn't match: %s != %s",
+                  id, value, pcmk__s(rule_input->rsc_agent, "unspecified"));
         return pcmk_rc_op_unsatisfied;
     }
 
+    crm_trace(PCMK_XE_RSC_EXPRESSION " %s passes (agent %s)",
```
Comment (nrwahl2) [inline lib/common/rules.c:1216]: I assume this is for convenience of reusing `value`. However, it is kind of odd if we have a rule that specifies standard and provider and we don't mention them here. We do give the rule ID so a reader can go look it up, but the agent name is no more special than the standard and provider names.

Up to you, it's just tracing

---

Also surprised to see that a totally empty resource expression is valid (no ID, no standard, no provider, no agent).

Maybe one more test for "empty resource expression matches"? All the individual missing cases are covered.

File: lib/common/rules.c [hunk 5]
```diff
@@ -1155,41 +1155,66 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
+         */
+        pcmk__config_warn(PCMK_XE_RSC_EXPRESSION " has no " PCMK_XA_ID);
+        id = "without ID"; // for logging
+    }
 
-    if (class != NULL &&
-        !pcmk__str_eq(class, rule_input->rsc_standard, pcmk__str_none)) {
-        crm_trace("Class doesn't match: %s != %s", class, rule_input->rsc_standard);
+    // Compare resource standard
+    value = crm_element_value(rsc_expression, PCMK_XA_CLASS);
```
Comment (nrwahl2) [inline lib/common/rules.c:1187]: Similar slight behavioral change to the one I mentioned for operation expressions:
Previously, `PCMK_XA_CLASS == "", rule_input->rsc_standard == "ocf"` would fail. Now, it passes. Same for provider and type.

Caveats similar to the ones I mentioned before also apply.

**Review by nrwahl2 — COMMENTED** [review]
Nice

File: lib/common/rules.c [hunk 6]
```diff
@@ -1155,41 +1155,77 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
+        crm_trace("Treating " PCMK_XE_OP_EXPRESSION " %s as not passing "
+                  "because no operation data was available", id);
         return pcmk_rc_op_unsatisfied;
     }
 
-    if ((interval_s != NULL) && (interval_ms != rule_input->op_interval_ms)) {
-        crm_trace("Interval doesn't match: %d != %d",
-                  interval_ms, rule_input->op_interval_ms);
+    // Compare operation interval (unspecified interval matches all)
+    if (!pcmk__str_empty(interval_s)
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c:1279]: `interval=""` does seem likely to be a user error, so keeping the current parse behavior and testing only NULL here makes sense.

File: lib/common/rules.c [hunk 7]
```diff
@@ -1155,41 +1155,77 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
+    // Compare operation name
     if (!pcmk__str_eq(name, rule_input->op_name, pcmk__str_none)) {
-        crm_trace("Name doesn't match: %s != %s", name, rule_input->op_name);
+        crm_trace(PCMK_XE_OP_EXPRESSION " %s fails because "
+                  "name doesn't match: %s != %s",
+                  id, name, rule_input->op_name);
         return pcmk_rc_op_unsatisfied;
     }
 
+    crm_trace(PCMK_XE_OP_EXPRESSION " %s passes (name %s, interval %s)",
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c:1295]: thx

File: lib/common/rules.c [hunk 8]
```diff
@@ -1155,41 +1155,77 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
+        /* @COMPAT When we can break behavioral backward compatibility,
+         * return pcmk_rc_op_unsatisfied
+         */
+        pcmk__config_warn(PCMK_XE_OP_EXPRESSION " element has no " PCMK_XA_ID);
+        id = "without ID"; // for logging
+    }
+
+    // Validate operation name
+    name = crm_element_value(op_expression, PCMK_XA_NAME);
+    if (pcmk__str_empty(name)) { // Not possible with schema validation enabled
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c:1256]: Research yielded some surprises.

An empty string here does pass schema validation. I expected that a required attribute must be non-empty, but RNG's `<text/>` (the default content for attributes) accepts empty strings. It should be possible to require non-empty strings or even matching a regular expression, using `<data type="string">` with extra parameters, but libxml2 appears to ignore parameters in a quick test.

Some RNG types (such as ID and `<value>`) normalize whitespace (stripping leading and trailing whitespace, and changing whitespace sequences to a single space) before validating. None of our code considers that. I'm not sure whether we should do anything about it now, but it does mean configurations can pass validation that will be rejected by the code.

Even before validation, XML parsers are supposed to normalize attribute values (https://www.w3.org/TR/xml11/#AVNormalize). libxml2 supposedly follows this, so I'm not sure why we don't see that.

Since empty strings are allowed here by RNG, I'm leaning to just checking NULL and letting the comparison fail.

File: lib/common/tests/rules/pcmk__evaluate_op_expression_test.c (hunk 1 above)
Comment (kgaillot) [reply lib/common/tests/rules/pcmk__evaluate_op_expression_test.c:109]: good idea

File: lib/common/tests/rules/pcmk__evaluate_op_expression_test.c (hunk 3 above)
Comment (kgaillot) [reply lib/common/tests/rules/pcmk__evaluate_op_expression_test.c:135]: thx

File: lib/common/rules.c (hunk 4 above)
Comment (kgaillot) [reply lib/common/rules.c:1216]: Yes, I didn't think it was worth separate variables just to trace them, but might as well

Comment (kgaillot) [reply lib/common/rules.c:1216]: It probably would have been a better design to require at least one of class, provider, and type, but it doesn't matter enough to change now

File: lib/common/rules.c (hunk 5 above)
Comment (kgaillot) [reply lib/common/rules.c:1187]: Thx

**Comment by kgaillot** [comment]
Updated per review

File: lib/common/rules.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: > Since empty strings are allowed here by RNG, I'm leaning to just checking NULL and letting the comparison fail.

Yeah I think that's what I would do. I don't care much, I just want to make you aware of any changes or edge cases I find so that you can make an informed decision.

And I know we can't make the schema tighter at-will.

File: lib/common/rules.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: That preserves existing behavior: `interval=""` would've previously failed to parse and then returned `FALSE`. My only real point is that the only possible `pcmk__str_empty()` case when we get here is `NULL`

Sounds good :)

File: lib/common/rules.c [hunk 9]
```diff
@@ -1155,41 +1155,70 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
     }
 
-    if ((interval_s != NULL) && (interval_ms != rule_input->op_interval_ms)) {
-        crm_trace("Interval doesn't match: %d != %d",
-                  interval_ms, rule_input->op_interval_ms);
+    // Compare operation name
+    if (!pcmk__str_eq(name, rule_input->op_name, pcmk__str_none)) {
+        crm_trace(PCMK_XE_OP_EXPRESSION " %s is unsatisfied because "
+                  "actual name '%s' doesn't match '%s'",
+                  id, pcmk__s(rule_input->op_name, ""), name);
```
Comment (nrwahl2) [inline lib/common/rules.c:1279]: This edgy of an edge case isn't worth worrying about, but: if `rule_input->op_name == NULL` and `name == ""`, this is going to say `actual name '' doesn't match ''`.

File: lib/common/rules.c [hunk 10]
```diff
@@ -1155,41 +1155,70 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
...
     }
 
-    if (!pcmk__str_eq(name, rule_input->op_name, pcmk__str_none)) {
-        crm_trace("Name doesn't match: %s != %s", name, rule_input->op_name);
+    // Compare operation interval (unspecified interval matches all)
+    if ((interval_s != NULL) && (interval_ms != rule_input->op_interval_ms)) {
+        crm_trace(PCMK_XE_OP_EXPRESSION " %s is unsatisfied because "
+                  "actual interval %s doesn't match %s",
+                  id, pcmk__readable_interval(interval_ms),
+                  pcmk__readable_interval(rule_input->op_interval_ms));
```
Comment (nrwahl2) [inline lib/common/rules.c:1288]: `rule_input->op_interval_ms` is the actual interval

File: lib/common/rules.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c]: I tried something like:

    <data type="string">
       <param name="minLength">1</param>
    </data>

which works with xmllint but appeared to be ignored when I tried it in the CIB schema (I didn't put a lot of time into it, so I may have messed it up). The `except` method doesn't rely on XSD datatypes so that may be better for that particular use case.

> Steps 1 through 3 of AV normalization apply,

Then why did we ever have to worry about newline escaping? Why didn't libxml2 provide us with already-escaped content?

[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: Ah okay. I've *only* tried with `xmllint` so far. Yeah `except` might work, no need to figure it out right now though.

---

> Why didn't libxml2 provide us with already-escaped content?

As best I remember[*] libxml2 simply doesn't do that for newlines. When serializing XML, it escapes characters that would make the output invalid, such as `'<'` and `'&'`, but not newlines.

[*] I use this qualifier because there are two libxml2 public API functions that are supposed to escape special characters, and IIRC there are more places internal to libxml2 that perform their own escaping... some of the behavior is questionable and it's hard to keep track of.

To further complicate things, we don't use libxml2 to serialize the whole XML document. In `pcmk__xml_string()` and its helpers, we dump the document piece-by-piece because we often need to filter out certain things, and libxml2 doesn't have that capability.

Concretely:
1. Set an attribute to `"some\nval"` (with literal newline). It's stored in memory as an `xmlAttr` with the literal newline.
2. Dump the XML document to text. Now there's a line break in the output:
```
test_attr="some
val"
```
3. Parse the XML document. Now the literal newline becomes a space. To prevent that from happening, the dumped text needs to have `test_attr="some&#x0A;val"`.

For a long time, we escaped all newlines as backslash-n when dumping them, so step 2's output had `test_attr="some\nval"`. Every future parse maintained and stored the literal backslash-n, and every future dump outputted it (without further escaping).

When we stopped escaping as backslash-n, we got the behavior demonstrated in the steps above.

Note that sending an IPC request dumps the request XML to text, and the IPC server parses it. So a sufficient condition for losing a newline is sending the XML via IPC. (There are other scenarios that involve dumping and parsing but that's a very common one.)

[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: > we don't use libxml2 to serialize the whole XML document

For a while we did, but only if `pcmk__xml_fmt_text` was set. That resulted in some output inconsistencies and prevented us from using const, so we dropped it.

There is one place (`pcmk__xml2fd()`) where we still use libxml2 to serialize the whole document. I've argued against keeping that before. We've dropped all but one caller (in `output_xml.c:xml_finish()`).
* I don't want to maintain it (though it is small)
* I worry about any inconsistencies that may arise compared to our usual custom dumping
* Any efficiency gain doesn't make a difference in `out->finish()`

File: lib/common/rules.c (hunk 9 above)
Comment (kgaillot) [reply lib/common/rules.c:1279]: Yeah I'm happy to ignore that for a trace log :)

File: lib/common/rules.c (hunk 10 above)
Comment (kgaillot) [reply lib/common/rules.c:1288]: Whoops

**Comment by kgaillot** [comment]
Updated to fix log mistake

**Review by nrwahl2 — APPROVED** [review]

File: lib/common/rules.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: Looking back, you were replying to "Steps 1 through 3 of AV normalization apply", so I might not have answered your question. Just in case:

If we're parsing a literal newline, we hit this:
> For a white space character (#x20, #xD, #xA, #x9), append a space character (#x20) to the normalized value.

If we're parsing `"&#x0A;"` (or equivalent), we hit this:
> For a character reference, append the referenced character to the normalized value.

This section doesn't talk about escaping but rather about how characters are parsed -- which differs depending on whether they were *already* escaped as character references or not

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c]: Right, so I would expect that libxml2 would replace newlines with spaces when the daemon parses the API request to set a multiline node attribute. But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.

[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: > But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.

It does. The point I'm making is that parsing XML from text is different from using `xmlSetProp()` or `crm_xml_add()` to set an attribute directly.

It normalizes newlines just fine when it's parsing them from text. If it's a literal newline, it converts it to a space. If it's a character reference, it preserves the newline. Demo (added to `crm_error.c` as a hack):
```
    {
        const char *s1 = "<n attr=\"a\nb\"/>";
        const char *s2 = "<n attr=\"a&#x0A;b\"/>";

        xmlNode *node = pcmk__xml_parse(s1);
        xmlAttr *attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");
        free_xml(node);

        node = pcmk__xml_parse(s2);
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");
        free_xml(node);
    }
```

Output:
```
$ crm_error asdf
a b
a
b
```

Note that the literal newline got parsed to a space, and the character reference was preserved as a newline.

---

By contrast, if I set the attribute directly (instead of parsing it from text), the literal newline is preserved. No need for a character reference. In fact, the character reference gets treated as a literal string instead of as a newline.
```
    {
        xmlNode *node = pcmk__xe_create(NULL, "n");
        xmlAttr *attr = NULL;

        xmlSetProp(node, (pcmkXmlStr) "attr", (pcmkXmlStr) "a\nb");
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");

        xmlSetProp(node, (pcmkXmlStr) "attr", (pcmkXmlStr) "a&#x0A;b");
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");

        free_xml(node);
    }
```

Output:
```
$ crm_error asdf
a
b
a&#x0A;b
```

[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: (ignore the weird "for" loop -- i initially had a more complicated printing scheme and then realized I didn't need it)

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/rules.c]: > > But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.
> 
> It does. The point I'm making is that parsing XML from text is different from using `xmlSetProp()` or `crm_xml_add()` to set an attribute directly.

Right, so why doesn't pacemaker-attrd normalize newlines when the attribute is parsed from the IPC XML, set the normalized value in its memory, and return that when queried?

[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/rules.c]: > Right, so why doesn't pacemaker-attrd normalize newlines when the attribute is parsed from the IPC XML, set the normalized value in its memory, and return that when queried?

It does. Before #3395 (note the newline is normalized as documented):
```
reid@laptop:~/git/pacemaker$ sudo attrd_updater -N laptop -n test_attr -U "$(echo -e "  a\nb  ")"
reid@laptop:~/git/pacemaker$ sudo attrd_updater -N laptop -n test_attr -Q
name="test_attr" host="laptop" value="  a b  "
```
